assets/app/view/game/spreadsheet.rb

```diff
@@ -337,8 +337,21 @@ def sorted_corporations
             @game.operating_order
           end
 
+        sort_order = @game.all_corporations.map { |c| [operating_order(c, operating_corporations), c] }
+                          .filter { |operating_order, _| operating_order }
+                          .sort_by! { |operating_order, _| operating_order }
+                          .map { |_, c| c }
+
         result = @game.all_corporations.map do |c|
-          operating_order = (operating_corporations.find_index(c) || -1) + 1
+          index = operating_corporations.find_index(c)
+          sort_order_key = sort_order.find_index(c)
+          operating_order = if index
+                              [2, index + 1]
+                            elsif sort_order_key
+                              [1, sort_order.length - sort_order_key]
+                            else
+                              [0, 0]
+                            end
```

Comment:
I found this code a little hard to read, after a bit of testing I think the following code accomplishes the same thing, and eliminates the need for the `operating_order()` function you added below:

```
        floated_order = 0
        unfloated_order = 0

        result = @game.all_corporations.sort.map do |c|
         operating_order =
           if operating_corporations[floated_order] == c
             floated_order += 1
             [FLOATED, floated_order]
           elsif c.ipoed
             unfloated_order += 1
             [UNFLOATED, unfloated_order]
           else
             [UNSTARTED, 0]
           end
          [operating_order, c]
        end
```

Up at the time you'll need to add these constants:

```
      FLOATED = 2
      UNFLOATED = 1
      UNSTARTED = 0
```

Would you mind testing that out to see if it gives you the same results?

Reply:
Thanks for taking a look - the result looks much cleaner :smile: 

There's a failing test case, because `@game.all_corporations.sort` throws an exception for minors - as they don't implement the compare function (this was covered by `sort_order_key = c.sort_order_key if c.respond_to?(:sort_order_key)`).

The `respond_to` is actually breaking the encapsulation, so it's hacky.

Reply:
Huh, looks like there are multiple ways different games handle the minors.

* [1822](https://github.com/tobymao/18xx/blob/ca01b6ef29aba08967ad0adf03f9915702a1d1b0/lib/engine/game/g_1822/entities.rb#L862-L864) and [1861](https://github.com/tobymao/18xx/blob/b3484429fc2a709d2a1b4aba60ffcc3b84f022fe/lib/engine/game/g_1861/game.rb#L911) - the minors are defined as `CORPORATIONS` and thus end up in `@corporations` instead of `@minors`
* [1873](https://github.com/tobymao/18xx/blob/b3484429fc2a709d2a1b4aba60ffcc3b84f022fe/lib/engine/game/g_1873/game.rb#L889-L891) and [1893](https://github.com/tobymao/18xx/blob/b3484429fc2a709d2a1b4aba60ffcc3b84f022fe/lib/engine/game/g_1893/game.rb#L745-L748) - `all_corporations` returns `@minors` + `@corporations`

I'll check out that failing test case, I'm thinking it make sense to implement `<=>` for minors like it's done on the [corporations](https://github.com/tobymao/18xx/blob/c15f2071c754697bbe905d4959889dc68d73a410/lib/engine/corporation.rb#L92-L105)

Reply:
Hmm thinking about that a little more it probably won't work, it would need the new `<=>` method to put things in the same order as `@game.operating_order` which doesn't really work.

Reply:
It actually does work - I've moved the spaceship operator into the operator module. WDYT?

Reply:
What you have here is pretty close to what I came up for that, even the new `1` at the start of the corporation array and the `0` for the minor. I don't really like magic numbers like those, and the corporation's `[1, 0, 0, 0, 0, name]` also made me wince when I typed that out.

I went back to your original `operating_order` function and `sort_order` array. I think with what I have here, `unfloated_corporations` is equivalent to your `sort_order` array. Thoughts?


```
        unfloated_corporations =
          (@game.all_corporations - operating_corporations)
            .filter { |c| c.respond_to?(:sort_order_key) && c.sort_order_key }
            .sort

        op_order = 0
        unfloated_order = 0

        result = @game.all_corporations.sort.map do |c|
          operating_order =
            if operating_corporations[op_order] == c
              op_order += 1
              [FLOATED, op_order]
            elsif unfloated_corporations[unfloated_order] == c
              unfloated_order += 1
              [UNFLOATED, unfloated_order]
            else
              [UNSTARTED, 0]
            end
          [operating_order, c]
        end
```

Reply:
Well I guess this isn't quite right either, it's failing the test `Assets #html 18MEX 17849 renders endgame` because some minors are being compared 🤔

Reply:
Oh I see, I still have a `sort` in the line starting with `result =`

Reply:
Ok think this is working now:

```
        unfloated_corporations =
          (@game.all_corporations - operating_corporations)
            .filter { |c| c.respond_to?(:sort_order_key) && c.sort_order_key }
            .sort

        op_order = 0
        unfloated_order = 0

        result = @game.all_corporations.map do |c|
          operating_order =
            if (index = operating_corporations.find_index(c))
              [FLOATED, index + 1]
            elsif (index = unfloated_corporations.find_index(c))
              [UNFLOATED, index + 1]
            else
              [UNSTARTED, 0]
            end
          [operating_order, c]
        end
```

Reply:
good idea :smile: